### PR TITLE
Add ContextManager

### PR DESCRIPTION
**Pull request by abitofevrything**
# Description

Moves context creation to its own class and cleans up `CommandsPlugin`.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] Ran `dart analyze .`
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: lib/src/context/context_manager.dart [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    IGuild? guild;
+    IMember? member;
+    IUser user;
+    if (message.guild != null) {
+      guild = await message.guild!.getOrDownload();
+
+      member = message.member;
+      user = await member!.user.getOrDownload();
+    } else {
+      user = message.author as IUser;
```
Comment (l7ssha) [inline lib/src/context/context_manager.dart:60]: rest of variables needs to be initialized

File: lib/src/context/context_manager.dart [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  ///
+  /// You might also be interested in:
+  /// - [createInteractionChatContext], for creating [IChatContext]s from interaction events.
+  Future<MessageChatContext> createMessageChatContext(
+    IMessage message,
+    StringView contentView,
+    String prefix,
+  ) async {
+    ChatCommand command =
+        commands.getCommand(contentView) ?? (throw CommandNotFoundException(contentView));
```
Comment (l7ssha) [inline lib/src/context/context_manager.dart:47]: brackets around throws are not needed here and I dont belive they make code easier to read

**Review by l7ssha — CHANGES_REQUESTED** [review]

File: lib/src/context/context_manager.dart (hunk 1 above)
Comment (abitofevrything) [reply lib/src/context/context_manager.dart:60]: No, guild and member should be left as null if the command was not run in a guild.

Comment (l7ssha) [reply lib/src/context/context_manager.dart:60]: then leaving them as Member? member initializes them to null by default?

Comment (abitofevrything) [reply lib/src/context/context_manager.dart:60]: Yes, nullable types default to null. Non nullable types are required to be initialised in all code paths before they are used.

Comment (l7ssha) [reply lib/src/context/context_manager.dart:60]: all right, then its good

**Review by l7ssha — APPROVED** [review]

**Pull request merged**